std/range/package.d

```diff
@@ -8647,3 +8651,269 @@ if (is(typeof(fun) == void) || isSomeFunction!fun)
+                foreach (i, _; Rs)
+                {
+                    case i:
+                        source[i].popFront();
+                        break;
+                }
+            }
+            _lastFrontIndex = frontIndex;
+        }
+
+        static if (isBidirectional)
+        {
+            @property auto ref back()
+            {
+                final switch (_lastBackIndex)
+                {
+                    foreach (i, _; Rs)
+                    {
+                        case i:
+                            assert(!source[i].empty);
+                            return source[i].back;
+                    }
+                }
+                assert(0);
+            }
+
+            private size_t backIndex()
+            {
+                import std.functional : binaryFun;
+                size_t bestIndex = size_t.max; // indicate undefined
+                E bestElement;
+                foreach (i, _; Rs)
+                {
+                    if (!source[i].empty)
+                    {
+                        if (bestIndex == size_t.max || // either this is the first or
+                            binaryFun!pred(bestElement, source[i].back))
+                        {
+                            bestIndex = i;
+                            bestElement = source[i].back;
+                        }
+                    }
+                }
+                return bestIndex;
+            }
+
+            void popBack()
+            {
+                final switch (_lastBackIndex)
+                {
+                    foreach (i, _; Rs)
+                    {
+                        case i:
+                            source[i].popBack();
+                            break;
+                    }
+                }
+                _lastBackIndex = backIndex;
+            }
+        }
+
+        static if (allSatisfy!(isForwardRange, staticMap!(Unqual, Rs)))
+        {
+            @property auto save()
+            {
+                Result result = this;
+                foreach (i, _; Rs)
+                {
+                    result.source[i] = result.source[i].save;
+                }
+                return result;
+            }
+        }
+
+        static if (allSatisfy!(hasLength, Rs))
+        {
+            @property size_t length()
+            {
+                size_t result;
+                foreach (i, _; Rs)
+                {
+                    result += source[i].length;
+                }
+                return result;
+            }
+
+            alias opDollar = length;
+        }
+
+        public Rs source;
+        private size_t _lastFrontIndex = size_t.max;
+        static if (isBidirectional)
+        {
+            private size_t _lastBackIndex = size_t.max;
+        }
+    }
+
+    return Result(rs);
+}
+
+@safe pure nothrow unittest
+{
+    import std.algorithm : equal;
+
+    alias S = short;
+    alias I = int;
+    alias D = double;
+
+    S[] a = [1, 2, 3];
```

Comment:
assumeSorted requires a range. If I try this

```D
    S[3] a = [1, 2, 3];
    // S[] a = [1, 2, 3];
    I[2] b = [50, 60];
    D[4] c = [10, 20, 30, 40];

    string[3] d = ["a", "b", "c"];

    static assert(!__traits(compiles, { auto m = merge(a.assumeSorted); }));
    static assert(!__traits(compiles, { auto m = merge(a.assumeSorted,
                                                       d.assumeSorted); }));

    auto m = merge(a[].assumeSorted,
                   b[].assumeSorted,
                   c[].assumeSorted);
```

it errors as

package.d(8893,23): Error: @nogc function 'std.range.__unittestL8874_180' cannot call non-@nogc function 'std.range.assumeSorted!("a < b", short[]).assumeSorted'
package.d(8894,23): Error: @nogc function 'std.range.__unittestL8874_180' cannot call non-@nogc function 'std.range.assumeSorted!("a < b", int[]).assumeSorted'
package.d(8895,23): Error: @nogc function 'std.range.__unittestL8874_180' cannot call non-@nogc function 'std.range.assumeSorted!("a < b", double[]).assumeSorted'